gnomad_qc/v4/annotations/generate_freq.py

```diff
@@ -461,90 +474,244 @@ def generate_faf_grpmax(ht: hl.Table) -> hl.Table:
     if pop_expr is not None:
-        scan_expr["pop_idx"] = hl.scan.counter(downsampling_ht.pop).get(
-            downsampling_ht.pop, 0
+        pop_counts = ht.aggregate(
+            hl.agg.filter(hl.is_defined(ht.pop), hl.agg.counter(ht.pop))
         )
+        downsamplings = [x for x in downsamplings if x <= sum(pop_counts.values())]
+        downsamplings = sorted(set(downsamplings + list(pop_counts.values())))
+        # Add an index by pop for use in computing frequencies, or other aggregate stats
+        # on the downsamplings.
+        scan_expr["pop_idx"] = hl.scan.counter(ht.pop).get(ht.pop, 0)
+    else:
+        pop_counts = None
+    logger.info("Found %i downsamplings: %s", len(downsamplings), downsamplings)
 
-    downsampling_ht = downsampling_ht.annotate(**scan_expr)
-    downsampling_ht = downsampling_ht.key_by("s").select(*scan_expr)
-    mt = mt.annotate_cols(downsampling=downsampling_ht[mt.s])
-    mt = mt.annotate_globals(
+    ht = ht.annotate(**scan_expr)
+    ht = ht.key_by("s").select(*scan_expr)
+
+    if isinstance(t, hl.MatrixTable):
+        t = t.annotate_cols(downsampling=ht[t.s])
+    else:
+        t = t.annotate(downsampling=ht[t.s])
+
+    t = t.annotate_globals(
         downsamplings=downsamplings,
-        ds_pop_counts=pop_counts if pop_expr is not None else None,
+        ds_pop_counts=pop_counts,
     )
 
-    return mt
+    return t
 
 
-def split_vds(
-    vds: hl.vds.VariantDataset, split_annotation: hl.expr.Expression
-) -> hl.ArrayExpression:
+# TODO: Move to gnomad_methods.
+def split_vds_by_strata(
+    vds: hl.vds.VariantDataset, strata_expr: hl.expr.Expression
+) -> List[hl.vds.VariantDataset]:
     """
-    Split a VDS into a list of VDSs based on the split_annotation.
+    Split a VDS into a list of VDSs based on `strata_expr`.
 
     :param vds: Input VDS.
-    :param split_annotation: Annotation on VDS variant_data MT split on.
+    :param strata_expr: Expression on VDS variant_data MT to split on.
     :return: List of VDSs.
     """
-    vds_list = []
-    for split in vds.variant_data.aggregate_cols(
-        hl.agg.collect_as_set(split_annotation)
-    ):
-        samples = vds.variant_data.filter_cols(split_annotation == split).s.collect()
-        vds_list.append(hl.vds.filter_samples(vds, samples))
-    return vds_list
-
-
-# The function below is an updated copy of Tim's freq updates found here:
-# https://github.com/broadinstitute/gnomad_methods/pull/537. I've merged the
-# additional_strata_expr and added the downsampling_expr functionality
-# that supports external downsampling annotations into this function
-def annotate_freq_and_high_ab_hets(
+    vmt = vds.variant_data
+    s_by_strata = vmt.aggregate_cols(
+        hl.agg.group_by(strata_expr, hl.agg.collect_as_set(vmt.s))
+    )
+
+    return [hl.vds.filter_samples(vds, list(s)) for strata, s in s_by_strata.items()]
+
+
+def compute_freq_by_strata(
```

Comment:
As is, it works like previously where some groupings of samples can have 0 samples. Do we want to remove these groups instead?